src/Microsoft.AspNetCore.Sockets.Common/Internal/Formatters/ServerSentEventsMessageParser.cs

```diff
@@ -0,0 +1,208 @@
+                        }
+                        break;
+                    case InternalParseState.ReadMessagePayload:
+                        if (!StartsWithDataPrefix(line))
+                        {
+                            throw new FormatException("Expected the message prefix 'data: '");
+                        }
+
+                        //Slice away the 'data: '
+                        var newData = line.Slice(line.IndexOf(ByteSpace) + 1).ToArray();
+                        start = lineEnd;
+                        _data.Add(newData);
+
+                        //Peek into next byte. If it is a carriage return byte, then advance to the next state
+                        if (reader.Peek() == ByteCR)
+                        {
+                            _internalParserState = InternalParseState.ReadEndOfMessage;
+                        }
+
+                        consumed = lineEnd;
+                        break;
+                    case InternalParseState.ReadEndOfMessage:
+                        if (ReadCursorOperations.Seek(start, end, out lineEnd, ByteLF) == -1)
+                        {
+                            // The message has ended with \r\n\r
+                            return ParseResult.Incomplete;
+                        }
+
+                        //To check for the invalid case of there being more data after the frame end \r\n\r\n
+                        if (!reader.End)
+                        {
+                            throw new FormatException("Unexpected data after line ending");
+                        }
+
+
+                        if (_data.Count > 0)
+                        {
+                            //Find the final size of the payload
+                            var payloadSize = 0;
+                            foreach (var dataLine in _data)
+                            {
+                                payloadSize += dataLine.Length;
+                            }
+                            var payload = new byte[payloadSize];
+
+                            //Copy the contents of the data array to a single buffer
+                            var marker = 0;
+                            foreach (var dataLine in _data)
+                            {
+                                dataLine.CopyTo(payload, marker);
+                                marker += dataLine.Length;
+                            }
+
+                            message = new Message(payload, messageType);
+                        }
+                        else
+                        {
+                            //Empty message
+                            message = new Message(new byte[0], messageType);
+                        }
+
+                        consumed = buffer.End;
+                        return ParseResult.Completed;
+                }
+            }
+            throw new FormatException("Error parsing data from event stream");
```

Comment:
This should just be `Incomplete`, right? We reached the end of our buffer but haven't seen a full frame yet.